Use pikepdf instead of PyPDF2, copy ToC over from original file

Kazy opened the pull request:
TLDR:
- First commit copy the ToC over from the original PDF. Disabled when the insertion of blank pages are enabled (because the ToC page numbers need to be recomputed), but it isn't difficult to implement.
- Second commit switches from `PyPDF2` to [pikepdf](https://pikepdf.readthedocs.io/en/latest/), gets the ToC issue fixed, an other issue with the ToC not showing up fixed, and much better performances.

This started as an attempt to keep the table of content after processing. The result is present in the first commit of this PR. 

While doing this, I noticed that some PDFs (like the one I was using for testing) had their table of content not recognized by my Remarkable 2. In my attempts at understanding why, I stumbled upon the `pikepdf` library and saw a line about the ability to `Repair, reformat or linearize PDFs`. I decided to try it by simply opening and saving the problematic PDF with the library and indeed it fixed the issue. So I modified the code to use `pikepdf` instead, which also solved the initial issue I wanted to solve (we never create a new PDF from scratch with `pikepdf` so the original ToC is kept) and brought better performances.

Two other reasons to stop using `PyPDF2`:
1. `PyPDF2` isn't maintained anymore.
2. On PDF files with lots of pages, you hit a recursion limit because of the way `PyPDF2` is written. The `_sweepIndirectReferences` function is recursive instead of being iterative.

One remark is that now when getting the raw bounding box with `pdftoppm`, a `Syntax Warning: Bad annotation destination` can appear, several times per page sometimes. I don't know what it means and everything seems to work fine.

Let me know what you think ! I've kept the first commit in case you don't want to switch to `pikepdf` but would still want to copy the ToC over.

Kazy commented:
Regarding the failing test, the string `arXiv:astro-ph/9207001v1  13 Jul 1992` seems to be present in the XML of the PDF and in the metadata at the end, but when looking at the PDF itself it isn't on the left of the first page anymore (if that's what the test is about).

GjjvdBurg commented:
Hi @Kazy,

Thanks so much for this contribution, this is awesome! I've looked into maintaining the ToC before, but never got far with the internals of PyPDF2. Swapping it out for pikepdf seems like a great solution.

I'll take a look at your PR right away.

GjjvdBurg commented on paper2remarkable/crop.py:
```diff
@@ -177,7 +174,12 @@ def get_raw_bbox_pdftoppm(self, filename, resolution=72):
+        # In case the process actually raises an exception, stderr can be recovered on the
+        # exception. See
+        # https://docs.python.org/3/library/subprocess.html#subprocess.CalledProcessError
+        im = subprocess.check_output(cmd, stderr=subprocess.PIPE)
```
Could we send stderr to ``subprocess.DEVNULL`` like is done in ``shrink_pdf``?

GjjvdBurg reviewed (commented):
Looks great, I had a minor suggestion regarding the syntax warning. Also, and this might be a slightly bigger change, but would it be possible to add a unit test for this?

GjjvdBurg commented:
> Regarding the failing test, the string `arXiv:astro-ph/9207001v1 13 Jul 1992` seems to be present in the XML of the PDF and in the metadata at the end, but when looking at the PDF itself it isn't on the left of the first page anymore (if that's what the test is about).

That's true, but it's strange that it's not caught by the filter. I'll take a look.

GjjvdBurg commented:
> One remark is that now when getting the raw bounding box with pdftoppm, a Syntax Warning: Bad annotation destination can appear, several times per page sometimes. I don't know what it means and everything seems to work fine.

What I suspect happens here is that because we extract a single page from the PDF when computing the bounding box, any links on that page will be invalid. Could it be that PyPDF2 always stripped the links, and that's why they only show up now?

Kazy commented:
> Could it be that PyPDF2 always stripped the links, and that's why they only show up now?

You were absolutely right. I removed the annotations from the page and the warnings disappeared.

I've also added a test to make sure the final PDF does have an outline. Let me know if you want something else tested, or if I should rewrite the current one in some other way !

GjjvdBurg commented:
Nice! Good to get rid of the warnings in this way. Thanks also for adding the test, always nice to have a unit test for functionality like this. One minor comment is that unless I'm mistaken it's not necessary to use the LocalFile provider in the test, it could just use the `Arxiv` provider, right?

GjjvdBurg commented:
I removed the failing test from master btw, it's indeed not necessary to have. So you should be able to merge master into your branch and the tests should pass.

Kazy commented:
> Nice! Good to get rid of the warnings in this way. Thanks also for adding the test, always nice to have a unit test for functionality like this. One minor comment is that unless I'm mistaken it's not necessary to use the LocalFile provider in the test, it could just use the `Arxiv` provider, right?

It could, I just thought it would be best to rely on less code (i.e. just the local filesystem instead of the Arxiv provider, which does more than just download the file from Arxiv) to test it. If you prefer I use the Arxiv provider I can change it though, no problem.

GjjvdBurg commented:
> It could, I just thought it would be best to rely on less code (i.e. just the local filesystem instead of the Arxiv provider, which does more than just download the file from Arxiv) to test it. If you prefer I use the Arxiv provider I can change it though, no problem.

That's a good point, but it actually suggests that doing _both_ might be even better, than we can be sure that one doesn't affect the other. I just tested it and it seems to work fine also with the arXiv provider, so I suggest we add a test with that one as well.

Would you mind adding that and merging master into your branch so that we have a passing test suite? Then I can merge the PR and make a release. Thank you!

Kazy commented:
Done and done !

GjjvdBurg commented:
Perfect! Thanks again for adding this, really appreciate it!
